booster/library/Booster/Pattern/Rewrite.hs

```diff
@@ -148,132 +178,196 @@ rewriteStep cutLabels terminalLabels pat = do
+                            logRemainder (map fst ruleApplicationData) SMT.IsUnsat groupRemainderPredicate
+                            pure $ mkBranch pat ruleApplicationData
+                        satRes@(SMT.IsSat{}) -> do
+                            -- the remainder condition is satisfiable.
+                            -- To construct the "remainder pattern",
+                            -- we add the remainder condition to the predicates of pat
+                            -- TODO: construct the remainder branch and consider it.
+                            throwRemainder (map fst ruleApplicationData) satRes groupRemainderPredicate
+                        satRes@SMT.IsUnknown{} -> do
+                            -- solver cannot solve the remainder
+                            -- TODO: descend into the remainder branch anyway, same as in the satisfiable case
+                            throwRemainder (map fst ruleApplicationData) satRes groupRemainderPredicate
+
+    labelOf = fromMaybe "" . (.ruleLabel) . (.attributes)
+    ruleLabelOrLocT = renderOneLineText . ruleLabelOrLoc
+    uniqueId = (.uniqueId) . (.attributes)
+
+    mkBranch ::
+        Pattern ->
+        [(RewriteRule "Rewrite", RewriteRuleAppliedData)] ->
+        RewriteResult Pattern
+    mkBranch base leafs =
+        RewriteBranch base $
+            NE.fromList $
+                map
+                    ( \(rule, RewriteRuleAppliedData{rewritten, rulePredicate, ruleSubstitution}) ->
+                        RewriteBranchNextState
+                            { ruleLabel = ruleLabelOrLocT rule
+                            , ruleUniqueId = uniqueId rule
+                            , rewrittenPat = rewritten
+                            , mRulePredicate = rulePredicate
+                            , ruleSubstitution
+                            }
+                    )
+                    leafs
+
+    -- check the remainder predicate for satisfiablity. Do nothing if unsat, abort rewriting otherwise
+    assertRemainderUnsat ::
+        LoggerMIO io => [RewriteRule "Rewrite"] -> Set.Set Predicate -> RewriteT io ()
+    assertRemainderUnsat rules remainderPrediate = do
+        isSatRemainder remainderPrediate >>= \case
+            SMT.IsUnsat -> pure ()
+            otherSatRes -> throwRemainder rules otherSatRes remainderPrediate
+
+    -- check the remainder predicate for satisfiability under the pre-branch pattern's constraints. Empty set is considered UNSAT
+    isSatRemainder :: LoggerMIO io => Set.Set Predicate -> RewriteT io (SMT.IsSatResult ())
+    isSatRemainder remainderPredicate =
+        if Set.null remainderPredicate || any isFalse remainderPredicate
+            then pure SMT.IsUnsat
+            else do
+                solver <- getSolver
+                SMT.isSat solver (Set.toList $ pat.constraints <> remainderPredicate) pat.substitution
```

Comment:
I find this `Empty set is considered UNSAT` very confusing!
The remainder is a set of predicates that is _and_ -ed for the check. If it is empty it should be SAT. Then of course we have to remove `assertRemainderUnsat` in the above code for "no rule applied".

Reply:
Very good point, I've removed the ` Set.null remainderPredicate` condition.

I've tweaked the code in "no rule applied" case to explicitly check that the remainder is empty and throw and exception if its not. I still want to guard against that since it is not obvious from the types that this case is impossible.